File: ios/calorietracker/Services/WorkoutFrameStore.swift
import CryptoKit
import Foundation
import os

/// Delivers authored workout frames without shipping the 1.2 GB corpus inside the IPA.
///
/// Resolution order for a frame:
/// 1. on-device cache (`Caches/WorkoutVectors/v2/<name>.<digest>.png`),
/// 2. an optional, gitignored developer sample folder (`calorietracker/WorkoutVectorsSample/`,
///    see scripts/build_workout_vectors_sample.py); release builds bundle no frames at all,
/// 3. download from the workout-vector CDN (`<base>/<name>.png?v=<digest>`), verified
///    against the manifest digest and stored in the cache.
///
/// Anything that fails resolves to nil and the UI keeps its placeholder, which is the
/// behaviour the app had before authored frames existed. Failures are remembered briefly
/// so an offline user does not retry every frame on every animation tick.
actor WorkoutFrameStore {
    nonisolated static let shared = WorkoutFrameStore()

    nonisolated static let defaultBaseURL = URL(string: "https://assets.fud-ai.app/workout-vectors/v2")!
    /// Debug-only override, e.g. launch argument `-WorkoutVectorsBaseURL http://localhost:8765`
    /// while `python3 -m http.server -d shared/workout-vectors 8765` serves the corpus.
    nonisolated static let baseURLOverrideKey = "WorkoutVectorsBaseURL"
    nonisolated static let bundledSampleDirectory = "WorkoutVectorsSample"

    nonisolated private static let failureRetryInterval: TimeInterval = 60
    nonisolated private static let maxFrameBytes = 4 * 1_024 * 1_024
    nonisolated private static let maxCacheBytes = 256 * 1_024 * 1_024
    nonisolated private static let trimTargetBytes = 192 * 1_024 * 1_024
    nonisolated private static let trimEveryDownloads = 16
    nonisolated private static let logger = Logger(subsystem: "com.apoorvdarshan.calorietracker", category: "WorkoutFrameStore")

    private let baseURL: URL?
    private let cacheDirectory: URL
    private let session: URLSession
    private var inFlight: [String: Task<URL?, Never>] = [:]
    private var recentFailures: [String: Date] = [:]
    private var downloadsSinceTrim = 0

    init(
        baseURL: URL? = WorkoutFrameStore.configuredBaseURL(),
        cacheDirectory: URL? = nil,
        session: URLSession = .shared
    ) {
        self.baseURL = baseURL
        self.cacheDirectory = cacheDirectory ?? Self.defaultCacheDirectory()
        self.session = session
    }

    // MARK: - Configuration

    nonisolated static func configuredBaseURL() -> URL? {
        #if DEBUG
        if
            let override = UserDefaults.standard.string(forKey: baseURLOverrideKey)?
                .trimmingCharacters(in: .whitespacesAndNewlines),
            !override.isEmpty
        {
            return override.lowercased() == "none" ? nil : URL(string: override)
        }
        #endif
        return defaultBaseURL
    }

    nonisolated static func defaultCacheDirectory() -> URL {
        let caches = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first
            ?? FileManager.default.temporaryDirectory
        return caches.appendingPathComponent("WorkoutVectors/v2", isDirectory: true)
    }

    // MARK: - Naming rules (shared with tests)

    nonisolated static func normalizedDigest(_ digest: String?) -> String? {
        guard let digest = digest?.lowercased(), (8...64).contains(digest.count) else { return nil }
        let hex = digest.unicodeScalars.allSatisfy { ("0"..."9").contains($0) || ("a"..."f").contains($0) }
        return hex ? digest : nil
    }

    nonisolated static func cacheFileName(for frame: ExerciseAuthoredFrame) -> String {
        "\(frame.name).\(normalizedDigest(frame.digest) ?? "nodigest").\(frame.fileExtension)"
    }

    nonisolated static func remoteURL(baseURL: URL?, frame: ExerciseAuthoredFrame) -> URL? {
        guard
            let baseURL,
            let scheme = baseURL.scheme?.lowercased(),
            scheme == "https" || scheme == "http",
            var components = URLComponents(
                url: baseURL.appendingPathComponent("\(frame.name).\(frame.fileExtension)"),
                resolvingAgainstBaseURL: false
            )
        else {
            return nil
        }
        if let digest = normalizedDigest(frame.digest) {
            components.queryItems = [URLQueryItem(name: "v", value: digest)]
        }
        return components.url
    }

    nonisolated static func data(_ data: Data, matchesDigest digest: String?) -> Bool {
        guard let expected = normalizedDigest(digest) else { return true }
        let actual = SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
        return actual.hasPrefix(expected)
    }

    nonisolated static func looksLikePNG(_ data: Data) -> Bool {
        let signature: [UInt8] = [0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A]
        return data.count > signature.count && Array(data.prefix(signature.count)) == signature
    }

    // MARK: - Lookup

    nonisolated func cacheFileURL(for frame: ExerciseAuthoredFrame) -> URL {
        cacheDirectory.appendingPathComponent(Self.cacheFileName(for: frame), isDirectory: false)
    }

    /// Frames from the gitignored `calorietracker/WorkoutVectorsSample/` developer folder.
    /// The synchronized Xcode group may copy it as a folder or flatten it, so check both.
    nonisolated static func bundledSampleURL(for frame: ExerciseAuthoredFrame) -> URL? {
        Bundle.main.url(
            forResource: frame.name,
            withExtension: frame.fileExtension,
            subdirectory: bundledSampleDirectory
        ) ?? Bundle.main.url(forResource: frame.name, withExtension: frame.fileExtension)
    }

    /// Frame file that can be shown without touching the network, if any.
    nonisolated func offlineURL(for frame: ExerciseAuthoredFrame) -> URL? {
        let cached = cacheFileURL(for: frame)
        if let size = try? cached.resourceValues(forKeys: [.fileSizeKey]).fileSize, size > 0 {
            return cached
        }
        return Self.bundledSampleURL(for: frame)
    }

    /// Local file URL for `frame`, downloading it first when needed. Nil when unavailable.
    func localURL(for frame: ExerciseAuthoredFrame) async -> URL? {
        if let url = offlineURL(for: frame) {
            return url
        }
        guard let remote = Self.remoteURL(baseURL: baseURL, frame: frame) else { return nil }
        if let failedAt = recentFailures[frame.name], Date().timeIntervalSince(failedAt) < Self.failureRetryInterval {
            return nil
        }
        if let task = inFlight[frame.name] {
            return await task.value
        }
        let task = Task<URL?, Never> { await self.download(frame, from: remote) }
        inFlight[frame.name] = task
        let result = await task.value
        inFlight[frame.name] = nil
        return result
    }

    /// Deletes every cached frame. Frames are re-downloadable, so this loses nothing.
    func clearCache() {
        try? FileManager.default.removeItem(at: cacheDirectory)
    }

    // MARK: - Download

    private func download(_ frame: ExerciseAuthoredFrame, from remote: URL) async -> URL? {
        let target = cacheFileURL(for: frame)
        do {
            var request = URLRequest(url: remote)
            request.setValue(frame.format == .svg ? "image/svg+xml" : "image/png", forHTTPHeaderField: "Accept")
            request.timeoutInterval = 30
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
                throw FrameError.badResponse
            }
            guard data.count <= Self.maxFrameBytes else { throw FrameError.tooLarge }
            guard frame.format != .png || Self.looksLikePNG(data) else { throw FrameError.notAnImage }
            guard Self.data(data, matchesDigest: frame.digest) else { throw FrameError.digestMismatch }

            let fileManager = FileManager.default
            try fileManager.createDirectory(at: cacheDirectory, withIntermediateDirectories: true)
            let partial = target.appendingPathExtension("part")
            try data.write(to: partial, options: .atomic)
            if fileManager.fileExists(atPath: target.path) {
                try fileManager.removeItem(at: target)
            }
            try fileManager.moveItem(at: partial, to: target)
            removeStaleRevisions(of: frame, keeping: target.lastPathComponent)
            recentFailures[frame.name] = nil
            downloadsSinceTrim += 1
            if downloadsSinceTrim % Self.trimEveryDownloads == 0 {
                trimCacheIfNeeded()
            }
            return target
        } catch {
            try? FileManager.default.removeItem(at: target.appendingPathExtension("part"))
            recentFailures[frame.name] = Date()
            Self.logger.notice("workout frame download failed: \(frame.name, privacy: .public): \(error.localizedDescription, privacy: .public)")
            return nil
        }
    }

    private func removeStaleRevisions(of frame: ExerciseAuthoredFrame, keeping keep: String) {
        let prefix = "\(frame.name)."
        guard let names = try? FileManager.default.contentsOfDirectory(atPath: cacheDirectory.path) else { return }
        for name in names where name.hasPrefix(prefix) && name != keep && !name.hasSuffix(".part") {
            try? FileManager.default.removeItem(at: cacheDirectory.appendingPathComponent(name))
        }
    }

    /// Keeps the frame cache bounded; the system may also purge Caches on its own.
    private func trimCacheIfNeeded() {
        let keys: Set<URLResourceKey> = [.fileSizeKey, .contentModificationDateKey]
        guard let urls = try? FileManager.default.contentsOfDirectory(
            at: cacheDirectory,
            includingPropertiesForKeys: Array(keys),
            options: [.skipsHiddenFiles]
        ) else {
            return
        }
        var entries: [(url: URL, size: Int, modified: Date)] = urls.compactMap { url in
            guard let values = try? url.resourceValues(forKeys: keys) else { return nil }
            return (url, values.fileSize ?? 0, values.contentModificationDate ?? .distantPast)
        }
        var total = entries.reduce(0) { $0 + $1.size }
        guard total > Self.maxCacheBytes else { return }
        entries.sort { $0.modified < $1.modified }
        for entry in entries {
            guard total > Self.trimTargetBytes else { break }
            if (try? FileManager.default.removeItem(at: entry.url)) != nil {
                total -= entry.size
            }
        }
    }

    nonisolated private enum FrameError: Error {
        case badResponse
        case tooLarge
        case notAnImage
        case digestMismatch
    }
}

File: android/app/src/main/java/com/apoorvdarshan/calorietracker/services/WorkoutFrameStore.kt
package com.apoorvdarshan.calorietracker.services

import android.annotation.SuppressLint
import android.content.Context
import android.os.SystemClock
import android.util.Log
import coil.ImageLoader
import coil.decode.DataSource
import coil.decode.ImageSource
import coil.decode.SvgDecoder
import coil.fetch.FetchResult
import coil.fetch.Fetcher
import coil.fetch.SourceResult
import coil.request.Options
import com.apoorvdarshan.calorietracker.BuildConfig
import com.apoorvdarshan.calorietracker.data.ExerciseVisualFormat
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import okio.Buffer
import okio.BufferedSource
import okio.Path.Companion.toOkioPath
import okio.buffer
import okio.source
import java.io.File
import java.io.IOException
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

/** One authored workout frame as referenced by the manifest. Coil request data for [WorkoutFrameFetcher]. */
data class WorkoutFrameRef(
    val name: String,
    val digest: String?,
    val format: ExerciseVisualFormat
) {
    val extension: String get() = if (format == ExerciseVisualFormat.SVG) "svg" else "png"
    val fileName: String get() = "$name.$extension"

    companion object {
        fun from(framePath: String, digest: String?, format: ExerciseVisualFormat): WorkoutFrameRef? {
            val name = WorkoutFrameLocator.frameName(framePath) ?: return null
            return WorkoutFrameRef(name, WorkoutFrameLocator.normalizedDigest(digest), format)
        }
    }
}

/**
 * Delivers authored workout frames without shipping the 1.2 GB corpus in the APK/AAB.
 *
 * Resolution order for a frame:
 * 1. on-device cache (`cacheDir/workout-vectors/v2/<name>.<digest>.png`),
 * 2. bundled asset (debug builds bundle the small sample pack; release bundles none),
 * 3. download from [BuildConfig.WORKOUT_VECTORS_BASE_URL] (`<base>/<name>.png?v=<digest>`),
 *    verified against the manifest digest and stored in the cache.
 *
 * Anything that fails resolves to null and the UI keeps its placeholder, which is the
 * same behaviour the app had before authored frames existed. Failures are remembered
 * briefly so offline users do not retry every frame on every recomposition.
 */
class WorkoutFrameStore private constructor(
    private val context: Context,
    private val baseUrl: String
) {
    private val cacheDirectory: File by lazy {
        File(context.cacheDir, CACHE_DIRECTORY).apply { mkdirs() }
    }
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val inFlight = ConcurrentHashMap<String, Deferred<File?>>()
    private val recentFailures = ConcurrentHashMap<String, Long>()
    /** Cache file names whose contents have been verified this process; avoids re-hashing per request. */
    private val verifiedFiles: MutableSet<String> = ConcurrentHashMap.newKeySet()
    private val downloadsSinceTrim = AtomicInteger(0)
    private val http: OkHttpClient by lazy {
        SecureHttpClient.builder()
            .connectTimeout(15, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .build()
    }

    /** Coil image loader that understands [WorkoutFrameRef] requests. */
    val imageLoader: ImageLoader by lazy {
        ImageLoader.Builder(context)
            .components {
                add(WorkoutFrameFetcher.Factory(this@WorkoutFrameStore))
                add(SvgDecoder.Factory())
            }
            .build()
    }

    internal class OpenedFrame(
        val imageSource: ImageSource,
        val dataSource: DataSource
    )

    /** Returns a readable source for [ref], downloading it first when needed, or null. */
    internal suspend fun open(ref: WorkoutFrameRef): OpenedFrame? = withContext(Dispatchers.IO) {
        cachedFile(ref)?.let { file ->
            return@withContext OpenedFrame(ImageSource(file.toOkioPath()), DataSource.DISK)
        }
        bundledAsset(ref)?.let { return@withContext it }
        val downloaded = ensureDownloaded(ref) ?: return@withContext null
        OpenedFrame(ImageSource(downloaded.toOkioPath()), DataSource.NETWORK)
    }

    /** True when the frame can be shown without touching the network. */
    fun isAvailableOffline(ref: WorkoutFrameRef): Boolean =
        cachedFile(ref) != null || hasBundledAsset(ref)

    /**
     * Returns the cached frame only if its contents verify against the manifest (size,
     * PNG signature and digest). Anything else is deleted so the next request repairs
     * it through the bundled or network path instead of failing to decode forever.
     */
    fun cachedFile(ref: WorkoutFrameRef): File? {
        val file = File(cacheDirectory, WorkoutFrameLocator.cacheFileName(ref.name, ref.digest, ref.extension))
        if (!file.isFile) return null
        if (file.name in verifiedFiles) return file
        val valid = file.length() in 1..MAX_FRAME_BYTES &&
            runCatching { verifyFrame(ref, file.readBytes()) }.isSuccess
        if (!valid) {
            verifiedFiles.remove(file.name)
            file.delete()
            return null
        }
        verifiedFiles.add(file.name)
        return file
    }

    private fun verifyFrame(ref: WorkoutFrameRef, bytes: ByteArray) {
        if (bytes.isEmpty() || bytes.size > MAX_FRAME_BYTES) throw IOException("Bad frame size: ${ref.name}")
        if (ref.extension == "png" && !WorkoutFrameLocator.looksLikePng(bytes)) {
            throw IOException("Not a PNG: ${ref.name}")
        }
        if (!WorkoutFrameLocator.matchesDigest(bytes, ref.digest)) {
            throw IOException("Digest mismatch for ${ref.name}")
        }
    }

    /** Reads at most [MAX_FRAME_BYTES]; aborts as soon as the response exceeds the cap. */
    private fun readBounded(source: BufferedSource, name: String): ByteArray {
        val buffer = Buffer()
        while (true) {
            val read = source.read(buffer, READ_CHUNK_BYTES)
            if (read == -1L) break
            if (buffer.size > MAX_FRAME_BYTES) throw IOException("Frame too large: $name")
        }
        return buffer.readByteArray()
    }

    private fun hasBundledAsset(ref: WorkoutFrameRef): Boolean =
        runCatching { context.assets.open(ref.fileName).close(); true }.getOrDefault(false)

    private fun bundledAsset(ref: WorkoutFrameRef): OpenedFrame? {
        val stream = runCatching { context.assets.open(ref.fileName) }.getOrNull() ?: return null
        return OpenedFrame(ImageSource(stream.source().buffer(), context), DataSource.DISK)
    }

    private suspend fun ensureDownloaded(ref: WorkoutFrameRef): File? {
        val url = WorkoutFrameLocator.remoteUrl(baseUrl, ref.name, ref.digest, ref.extension) ?: return null
        val failedAt = recentFailures[ref.name]
        if (failedAt != null && SystemClock.elapsedRealtime() - failedAt < FAILURE_RETRY_MS) return null

        val deferred = inFlight.getOrPut(ref.name) {
            scope.async { download(ref, url) }
        }
        return try {
            deferred.await()
        } finally {
            inFlight.remove(ref.name, deferred)
        }
    }

    private fun download(ref: WorkoutFrameRef, url: String): File? {
        cachedFile(ref)?.let { return it }
        val target = File(cacheDirectory, WorkoutFrameLocator.cacheFileName(ref.name, ref.digest, ref.extension))
        val partial = File(cacheDirectory, target.name + ".part")
        return try {
            val request = Request.Builder().url(url).header("Accept", "image/png").get().build()
            val bytes = http.newCall(request).execute().use { response ->
                if (!response.isSuccessful) throw IOException("HTTP ${response.code} for ${ref.name}")
                val body = response.body ?: throw IOException("Empty body for ${ref.name}")
                // Early rejection only; the cap is enforced while streaming because the
                // declared length may be absent (chunked) or wrong.
                if (body.contentLength() > MAX_FRAME_BYTES) throw IOException("Frame too large: ${ref.name}")
                readBounded(body.source(), ref.name)
            }
            verifyFrame(ref, bytes)
            cacheDirectory.mkdirs()
            partial.writeBytes(bytes)
            verifiedFiles.remove(target.name)
            if (!partial.renameTo(target)) {
                // Non-atomic fallback; cachedFile() re-verifies so an interrupted copy self-heals.
                partial.copyTo(target, overwrite = true)
                partial.delete()
            }
            verifiedFiles.add(target.name)
            removeStaleRevisions(ref, keep = target.name)
            recentFailures.remove(ref.name)
            if (downloadsSinceTrim.incrementAndGet() % TRIM_EVERY_DOWNLOADS == 0) trimCache()
            target
        } catch (error: Exception) {
            partial.delete()
            recentFailures[ref.name] = SystemClock.elapsedRealtime()
            Log.w(TAG, "workout frame download failed: ${ref.name}: ${error.message}")
            null
        }
    }

    private fun removeStaleRevisions(ref: WorkoutFrameRef, keep: String) {
        val prefix = WorkoutFrameLocator.cacheFilePrefix(ref.name)
        cacheDirectory.listFiles()
            ?.filter { it.name.startsWith(prefix) && it.name != keep && !it.name.endsWith(".part") }
            ?.forEach {
                verifiedFiles.remove(it.name)
                it.delete()
            }
    }

    /** Keeps the frame cache bounded; frames are re-downloadable so eviction is harmless. */
    private fun trimCache() {
        val files = cacheDirectory.listFiles()?.filter { it.isFile } ?: return
        var total = files.sumOf { it.length() }
        if (total <= MAX_CACHE_BYTES) return
        for (file in files.sortedBy { it.lastModified() }) {
            if (total <= TRIM_TARGET_BYTES) break
            val size = file.length()
            if (file.delete()) {
                verifiedFiles.remove(file.name)
                total -= size
            }
        }
    }

    /** Deletes every cached frame (Settings → storage management, tests). */
    fun clearCache() {
        verifiedFiles.clear()
        cacheDirectory.listFiles()?.forEach { it.delete() }
    }

    companion object {
        private const val TAG = "WorkoutFrameStore"
        private const val CACHE_DIRECTORY = "workout-vectors/v2"
        private const val FAILURE_RETRY_MS = 60_000L
        private const val MAX_FRAME_BYTES = 4L * 1024 * 1024
        private const val READ_CHUNK_BYTES = 64L * 1024
        private const val MAX_CACHE_BYTES = 256L * 1024 * 1024
        private const val TRIM_TARGET_BYTES = 192L * 1024 * 1024
        private const val TRIM_EVERY_DOWNLOADS = 16

        // Holds the application context only (see get()), never an Activity.
        @SuppressLint("StaticFieldLeak")
        @Volatile
        private var instance: WorkoutFrameStore? = null

        fun get(context: Context): WorkoutFrameStore =
            instance ?: synchronized(this) {
                instance ?: WorkoutFrameStore(
                    context.applicationContext,
                    BuildConfig.WORKOUT_VECTORS_BASE_URL
                ).also { instance = it }
            }
    }
}

/** Coil fetcher that resolves [WorkoutFrameRef] through [WorkoutFrameStore]. */
internal class WorkoutFrameFetcher(
    private val ref: WorkoutFrameRef,
    private val store: WorkoutFrameStore
) : Fetcher {
    override suspend fun fetch(): FetchResult {
        val opened = store.open(ref) ?: throw IOException("Workout frame unavailable: ${ref.name}")
        return SourceResult(
            source = opened.imageSource,
            mimeType = if (ref.extension == "svg") "image/svg+xml" else "image/png",
            dataSource = opened.dataSource
        )
    }

    class Factory(private val store: WorkoutFrameStore) : Fetcher.Factory<WorkoutFrameRef> {
        override fun create(data: WorkoutFrameRef, options: Options, imageLoader: ImageLoader): Fetcher =
            WorkoutFrameFetcher(data, store)
    }
}
